pyadjoint/optimization/tao_solver.py

```diff
@@ -58,6 +55,7 @@ def __init__(self, X, *, comm=None):
         self._indices = tuple(indices)
         self._n = n
         self._N = N
+        self._index_concatenated_vecs = None
```

Comment:
I think this should be initialized using an initial `VecConcatenate`. `self._indices` is then not needed. This does assume that `VecConcatenate` has predictable behaviour so that we can reuse the `IS` objects even after concatenating different vectors with the same shape.